File: core/TrappingRainWater/TrappingRainWater.java
package com.dbc;

import java.util.Arrays;

public class TrappingRainWater {
    public int trap(int[] height) {
        int[] leftDp = new int[height.length];
        Arrays.fill(leftDp, 0);
        int[] rightDp = new int[height.length];
        Arrays.fill(rightDp, 0);
        for (int i = 1; i < height.length; i++) {
            leftDp[i] = Math.max(leftDp[i - 1], height[i - 1]);
        }
        for (int i = height.length - 2; i >= 0; i--) {
            rightDp[i] = Math.max(rightDp[i + 1], height[i + 1]);
        }
        int res = 0;
        for (int i = 1; i < height.length - 1; i++) {
            if (leftDp[i] > height[i] && rightDp[i] > height[i]) {
                res += Math.min(leftDp[i], rightDp[i]) - height[i];
            }
        }
        return res;
    }
}

File: core/TrappingRainWater/TrappingRainWater.py
from typing import List


def trap(height: List[int]) -> int:
    length = len(height)
    left_dp, right_dp = [0] * length, [0] * length
    for index in range(1, length):
        left_dp[index] = max(left_dp[index - 1], height[index - 1])
    for index in range(length - 2, -1, -1):
        right_dp[index] = max(right_dp[index + 1], height[index + 1])
    res = 0
    for index in range(1, length - 1):
        if left_dp[index] > height[index] and right_dp[index] > height[index]:
            res += min(left_dp[index], right_dp[index]) - height[index]

    return res


